File: crates/dux-web/web/src/lib/flatList.ts
// Pure helpers backing the flat agent list, shared by the desktop sidebar and
// the mobile hub so the two surfaces never drift. Kept free of React and
// dnd-kit so every rule here is unit-testable.
//
// The ordering (main-bucket sort plus the Active-mode recency-sorted quiet
// tail) is a twin of the core-owned `dux_core::flat_list::order_sessions`,
// pinned by shared vectors.

import { sortedSessionIds, type SortKey as BaseSortKey } from "@/lib/sortSessions"
import type { SessionView } from "@/lib/types"

// The flat list's display sort. "active" is the default (core order with
// working / needs-attention agents floated to the top); the remaining keys
// reuse the TUI-parity comparators; "manual" is the raw persisted order.
export type FlatSortKey = "active" | BaseSortKey | "manual"

export const FLAT_SORT_LABELS: Record<FlatSortKey, string> = {
  active: "Active first",
  updated: "Recently updated",
  created: "Recently created",
  name: "Name (A to Z)",
  // The web picker never offers name_desc (only the TUI cycles into it) but has
  // to display it when the TUI set it: the menu appends a checked row for it
  // while it is active, since the trigger always reads "Sort".
  name_desc: "Name (Z to A)",
  manual: "Manual order",
}

// A session is "quiet" when it is not active (detached or exited): dormant work
// that collapses into the Quiet tail instead of hogging the main list.
export function isQuietSession(session: SessionView): boolean {
  return session.status !== "active"
}

// Whether the Quiet tail renders forced-open for the current search. Twin of
// the core-owned `dux_core::quiet_tail::quiet_tail_forced_open`, pinned by
// shared vectors. Both queries are trimmed and lowercased (empty means no
// search, `dismissedQuery` is the one the user last collapsed the tail under),
// so a whitespace or case variant does not resurrect a dismissed tail.
export function quietTailForcedOpen(
  normalizedQuery: string,
  dismissedQuery: string | null,
  hasQuietHit: boolean,
): boolean {
  if (normalizedQuery === "") return false
  if (dismissedQuery === normalizedQuery) return false
  return hasQuietHit
}

// Split the sessions into the always-visible MAIN list (active agents) and the
// collapsible QUIET tail (detached / exited). Input order is preserved within
// each bucket so a later sort/partition sees a stable starting order.
export function partitionQuiet(sessions: SessionView[]): {
  main: SessionView[]
  quiet: SessionView[]
} {
  const main: SessionView[] = []
  const quiet: SessionView[] = []
  for (const session of sessions) {
    if (isQuietSession(session)) quiet.push(session)
    else main.push(session)
  }
  return { main, quiet }
}

// Stable active-first ordering: agents that are working or need attention rise
// above the rest, each group keeping the caller's incoming (core) order. This is
// a lightweight float, not a full re-sort, the underlying order is untouched, so
// it composes cleanly over the persisted drag order.
export function activeFirstSessions(sessions: SessionView[]): SessionView[] {
  const hot: SessionView[] = []
  const rest: SessionView[] = []
  for (const session of sessions) {
    if (session.working || session.needs_attention) hot.push(session)
    else rest.push(session)
  }
  return [...hot, ...rest]
}

// Order the MAIN sessions for display by the chosen sort. "active" floats the hot
// agents up (default); "manual" is the caller's order verbatim (the persisted
// drag order); the remaining keys reuse `sortedSessionIds` so the flat list sorts
// identically to the TUI palette commands.
export function sortMainSessions(
  sessions: SessionView[],
  key: FlatSortKey,
): SessionView[] {
  if (key === "active") return activeFirstSessions(sessions)
  if (key === "manual") return sessions.slice()
  const order = sortedSessionIds(sessions, key as BaseSortKey)
  const byId = new Map(sessions.map((session) => [session.id, session]))
  return order
    .map((id) => byId.get(id))
    .filter((session): session is SessionView => session !== undefined)
}

// Order the quiet (inactive) tail for display: in "active" mode most recently
// updated first, every other mode verbatim, matching the core-owned
// `flat_list::order_sessions` so the surfaces agree.
export function sortQuietTail(
  sessions: SessionView[],
  key: FlatSortKey,
): SessionView[] {
  if (key !== "active") return sessions.slice()
  const order = sortedSessionIds(sessions, "updated")
  const byId = new Map(sessions.map((session) => [session.id, session]))
  return order
    .map((id) => byId.get(id))
    .filter((session): session is SessionView => session !== undefined)
}

// The agent to land on when the focused one vanishes. "Next" is read off the
// ordering the user is looking at, the main bucket in the current sort mode:
// `previous` is the list as it was while the gone agent still existed, which
// gives next a position to count from, and `current` is the list that just
// arrived. The scan wraps, so deleting the last row lands on the first. Only
// active agents are candidates, since a dormant one has no process to land in;
// null when none is left, which the caller renders as home.
export function nextActiveSessionId(
  previous: SessionView[],
  current: SessionView[],
  goneSessionId: string,
  key: FlatSortKey,
): string | null {
  const candidates = sortMainSessions(partitionQuiet(current).main, key).map(
    (session) => session.id,
  )
  if (candidates.length === 0) return null
  const candidateSet = new Set(candidates)
  const before = sortMainSessions(partitionQuiet(previous).main, key).map(
    (session) => session.id,
  )
  const at = before.indexOf(goneSessionId)
  // The gone agent was not in the active bucket (it was quiet, or this client
  // never saw it): there is no position to count from, so take the first row.
  if (at === -1) return candidates[0]
  for (let step = 1; step <= before.length; step++) {
    const id = before[(at + step) % before.length]
    if (candidateSet.has(id)) return id
  }
  // Every agent that shared the old active bucket is gone too; fall back to
  // whatever the new list starts with.
  return candidates[0]
}

// Agent order is one global flat list, persisted by `reorderAgents` as a plain
// move over every session id.

// The drag baseline for a drop: the complete session id list in the order the
// user is looking at. A drop made in a computed mode (active/name/updated/
// created) totalizes what the screen showed, the main list then the quiet tail
// below it. Every session is included, never the visible or filtered subset,
// because the persisted order is total. "manual" is deliberately the base order
// verbatim, quiet sessions interleaved where the base has them.
export function displayedSessionOrder(
  sessions: SessionView[],
  key: FlatSortKey,
): string[] {
  if (key === "manual") return sessions.map((session) => session.id)
  const { main, quiet } = partitionQuiet(sessions)
  return [...sortMainSessions(main, key), ...sortQuietTail(quiet, key)].map(
    (session) => session.id,
  )
}

// The colored state word on a row's second line, read off the same flags that
// drive the working pulse and the attention pulse so the word and the motion
// cue cannot disagree. Colors are Tailwind palette utilities, as in agentRow.ts, never raw
// hex or oklch.
export interface StateWord {
  label: string
  className: string
}

export function stateWord(session: SessionView): StateWord {
  // The priority ladder is the core-owned `dux_core::row_state::agent_row_state`
  // decision, pinned by shared vectors; this surface only words and colors it.
  if (session.needs_attention) return { label: "Needs you", className: "text-cyan-100" }
  if (session.status === "active" && session.typing) {
    // The soft-violet typing token, matching the TUI's `#c586e0` typing hue.
    return { label: "Typing", className: "text-dux-typing" }
  }
  if (session.status === "active" && session.working) {
    // Match the app's active status color (agentRow.ts STATUS_DOT_COLOR),
    // not a new palette hue.
    return { label: "Working", className: "text-green-500" }
  }
  if (session.status === "active") {
    return { label: "Idle", className: "text-muted-foreground" }
  }
  if (session.status === "detached") {
    return { label: "Detached", className: "text-amber-500" }
  }
  return { label: "Exited", className: "text-muted-foreground" }
}

File: crates/dux-core/src/flat_list.rs
//! The flat agent-list ORDERING decision, core-owned and shared by rule with the
//! web's `crates/dux-web/web/src/lib/flatList.ts` + `sortSessions.ts`. This is a
//! cross-language twin: the rule lives here in Rust (the source of truth), the TS
//! keeps a hand-written mirror, and the two are pinned by SHARED TEST VECTORS
//! (the cases below are duplicated in `flatList.test.ts` / `sortSessions.test.ts`,
//! in the `agent_search.rs` / `agentSearch.ts` style) so a comparator change in
//! one language that is not mirrored fails a test.
//!
//! The list is a single globally-ordered agent list with no project grouping.
//! Active agents form the MAIN bucket; detached/exited ("inactive"/"quiet") agents
//! form a collapsible TAIL. This module owns only the ORDERING (which indices, in
//! which order); each surface wraps the result into its own list-item type.

use crate::model::{AgentSession, SessionStatus};

/// The display sort applied to the flat list. Mirrors the TUI's `AgentSortMode`
/// and the web's `FlatSortKey`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FlatSortMode {
    /// Working / needs-attention agents float to the top (a stable float, not a
    /// re-sort); the inactive tail sorts most-recently-active-first. The default.
    Active,
    /// Most recently updated first (`Reverse(updated_at)`).
    Updated,
    /// Most recently created first (`Reverse(created_at)`).
    Created,
    /// By name (title-or-branch, case-insensitive), A to Z.
    NameAsc,
    /// By name (title-or-branch, case-insensitive), Z to A.
    NameDesc,
    /// The stored global order verbatim (the web's drag-reorder order).
    Manual,
}

/// The ordered result: main (active) indices, then the inactive/quiet tail
/// indices. Indices point into the input `sessions` slice.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FlatOrder {
    pub active: Vec<usize>,
    pub inactive: Vec<usize>,
}

/// Order the sessions for the flat list. `is_hot(index)` reports whether the
/// session is working or needs attention (used only by `Active`'s float).
/// `is_visible(index)` is a symmetric DISPLAY filter: an index that fails it never
/// enters either bucket. `sessions` is never mutated.
///
/// The partition always preserves incoming order first, then each bucket is
/// ordered by `mode`:
/// - `Active`: the main bucket floats `is_hot` indices up (a STABLE float keeping
///   incoming order within each group); the inactive tail sorts by most recently
///   updated (`Reverse(updated_at)`), most-recently-active-first.
/// - `Updated` / `Created` / `NameAsc` / `NameDesc`: the MAIN bucket is sorted by
///   that comparator; the inactive tail stays VERBATIM (only Active reorders the
///   tail, so the two surfaces agree).
/// - `Manual`: both buckets verbatim (the stored global order).
pub fn order_sessions(
    sessions: &[AgentSession],
    mode: FlatSortMode,
    is_hot: &dyn Fn(usize) -> bool,
    is_visible: &dyn Fn(usize) -> bool,
) -> FlatOrder {
    let mut active: Vec<usize> = Vec::new();
    let mut inactive: Vec<usize> = Vec::new();
    for (index, session) in sessions.iter().enumerate() {
        if !is_visible(index) {
            continue;
        }
        match session.status {
            SessionStatus::Detached | SessionStatus::Exited => inactive.push(index),
            _ => active.push(index),
        }
    }

    // Case-insensitive name key: title-or-branch lowercased. Rust's `str::cmp` on
    // this lowercased key matches the web's code-point comparison in
    // `sortSessions.ts` (UTF-8 byte order equals code-point order).
    let name_key = |index: usize| -> String {
        let s = &sessions[index];
        s.display_label().to_lowercase()
    };
    // Comparator-based ordering for the MAIN bucket in the non-Active modes.
    // `sort_by_key` is stable, so equal keys keep incoming order.
    let order_bucket = |bucket: &mut Vec<usize>| match mode {
        FlatSortMode::Updated => bucket.sort_by_key(|&i| std::cmp::Reverse(sessions[i].updated_at)),
        FlatSortMode::Created => bucket.sort_by_key(|&i| std::cmp::Reverse(sessions[i].created_at)),
        FlatSortMode::NameAsc => bucket.sort_by_key(|&i| name_key(i)),
        FlatSortMode::NameDesc => bucket.sort_by_key(|&i| std::cmp::Reverse(name_key(i))),
        FlatSortMode::Active | FlatSortMode::Manual => {}
    };

    match mode {
        FlatSortMode::Active => {
            // Stable float: hot indices rise above the rest, each group keeping
            // incoming order.
            let mut hot: Vec<usize> = Vec::new();
            let mut rest: Vec<usize> = Vec::new();
            for &i in &active {
                if is_hot(i) {
                    hot.push(i);
                } else {
                    rest.push(i);
                }
            }
            active = hot;
            active.extend(rest);
            // The collapsed tail sorts most-recently-active-first.
            inactive.sort_by_key(|&i| std::cmp::Reverse(sessions[i].updated_at));
        }
        FlatSortMode::Manual => {
            // Both buckets stay verbatim (incoming order).
        }
        _ => {
            // Only the MAIN bucket sorts by the comparator; the inactive tail stays
            // verbatim (only Active reorders the tail).
            order_bucket(&mut active);
        }
    }

    FlatOrder { active, inactive }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::test_support::sample_session;
    use chrono::{TimeZone, Utc};

    fn session(id: &str, status: SessionStatus, updated_hour: u32, name: &str) -> AgentSession {
        let mut s = sample_session(id, "p1", name);
        s.status = status;
        s.title = Some(name.to_string());
        s.updated_at = Utc
            .with_ymd_and_hms(2026, 7, 17, updated_hour, 0, 0)
            .unwrap();
        s.created_at = Utc
            .with_ymd_and_hms(2026, 7, 17, updated_hour, 0, 0)
            .unwrap();
        s
    }

    // ── SHARED VECTORS with flatList.test.ts ───────────────────────────────────

    #[test]
    fn active_mode_floats_hot_and_recency_sorts_the_tail() {
        // main: zeta(active), alpha(active), hot(active+hot); tail: gone(exited),
        // parked(detached) with distinct updated_at.
        let sessions = vec![
            session("zeta", SessionStatus::Active, 12, "zeta"),
            session("gone", SessionStatus::Exited, 10, "gone"),
            session("alpha", SessionStatus::Active, 12, "alpha"),
            session("hot", SessionStatus::Active, 12, "hot"),
            session("parked", SessionStatus::Detached, 14, "parked"),
        ];
        let is_hot = |i: usize| sessions[i].id == "hot";
        let order = order_sessions(&sessions, FlatSortMode::Active, &is_hot, &|_| true);
        // Hot floats up, the rest keep incoming order.
        assert_eq!(order.active, vec![3, 0, 2]); // hot, zeta, alpha
        // Tail is most-recently-updated first: parked(14) before gone(10).
        assert_eq!(order.inactive, vec![4, 1]); // parked, gone
    }

    #[test]
    fn non_active_modes_leave_the_tail_verbatim() {
        let sessions = vec![
            session("old", SessionStatus::Detached, 10, "old"),
            session("newest", SessionStatus::Exited, 14, "newest"),
            session("mid", SessionStatus::Detached, 12, "mid"),
        ];
        // Name mode: no active rows, tail stays in incoming order (old, newest, mid).
        let order = order_sessions(&sessions, FlatSortMode::NameAsc, &|_| false, &|_| true);
        assert!(order.active.is_empty());
        assert_eq!(order.inactive, vec![0, 1, 2]);
    }

    #[test]
    fn name_mode_sorts_the_main_bucket_only() {
        let sessions = vec![
            session("b", SessionStatus::Active, 12, "bravo"),
            session("a", SessionStatus::Active, 12, "alpha"),
        ];
        let order = order_sessions(&sessions, FlatSortMode::NameAsc, &|_| false, &|_| true);
        assert_eq!(order.active, vec![1, 0]); // alpha, bravo
    }

    #[test]
    fn manual_keeps_both_buckets_verbatim() {
        let sessions = vec![
            session("z", SessionStatus::Active, 12, "z"),
            session("gone", SessionStatus::Exited, 10, "gone"),
            session("a", SessionStatus::Active, 12, "a"),
            session("parked", SessionStatus::Detached, 14, "parked"),
        ];
        let order = order_sessions(&sessions, FlatSortMode::Manual, &|_| false, &|_| true);
        assert_eq!(order.active, vec![0, 2]);
        assert_eq!(order.inactive, vec![1, 3]); // verbatim, NOT recency-sorted
    }

    #[test]
    fn invisible_indices_enter_neither_bucket() {
        let sessions = vec![
            session("a", SessionStatus::Active, 12, "a"),
            session("b", SessionStatus::Active, 12, "b"),
        ];
        let order = order_sessions(&sessions, FlatSortMode::Active, &|_| false, &|i| {
            sessions[i].id == "a"
        });
        assert_eq!(order.active, vec![0]);
    }
}
